docs/src/design/many_differentials.md

```diff
@@ -0,0 +1,188 @@
+```julia
+julia> struct TimeSample
+           time::DateTime
+           value::Float64
+       end
+```
+
+We can look at its structure:
+```julia
+julia> dump(TimeSample(now(), 2.6e9))
+TimeSample
+  time: DateTime
+    instant: Dates.UTInstant{Millisecond}
+      periods: Millisecond
+        value: Int64 63720043490844
+  value: Float64 2.6e9
+```
+
+Thus we see the that structural differential would be:
+```julia
+Composite{TimeSample}(
+    time::Composite{DateTime}(
+        instant::Composite{UTInstant{Millisecond}}(
+            periods::Composite{Millisecond}(
+                value::Int64
+            )
+        )
+    ),
+    value::Float64
+)
+```
+
+But instead in the custom sensitivity rule we would write a semi-structured differential type.
+Since there is not a natural differential type for `TimeSample` but there is for `DateTime`.
+```julia
+Composite{TimeSample}(
+    time::Day,
+    value::Float64
+)
+```
+
+So the rule author has written a structural differential with some fields that are natural differentials.
+
+Another related case is for types that overload `getproperty` such as `SVD` and `QR`.
+In this case the structural differential will be based on the fields, but those fields do not always have an easy relation to what is actually used in math.
+For example, the `QR` type has fields `factors` and `t`, but we would more naturally think in terms of the properties `Q` and `R`.
+So most rule authors would want to write semi-structural differentials based on the properties.
+
+To return to the question of why ChainRules has `Composite{P, <:NamedTuple}` whereas Zygote v0.4 just has `NamedTuple`, it relates to semi-structural derivatives, and being able to overload things more generally.
+If one knows that one has a semi-structural derivative based on property names, like `Composite{QR}(Q=..., R=...)`, and one is adding it to the true structural derivative based on field names `Composite{QR}(factors=..., τ=...)`, then we need to overload the addition operator to perform that correctly.
+We cannot overload happily similar things for `NamedTuple` since we don't know the primal type, only the names of the values contained.
+In fact we can't actually overload addition at all for `NamedTuple` as that would be type-piracy, so have to use `Zygote.accum` instead.
+
+Another use of the primal being a type parameter is to catch errors.
+ChainRules disallows the addition of `Composite{SVD}` to `Composite{QR}` since in a correctly differentiated program that can never occur.
+
+## Differentials types for computational efficiency
+
+There is another kind of unnatural differential.
+One that is for computational efficiency.
+ChainRules has [`Thunk`](@ref)s and [`InplaceableThunk`](@ref)s, which wrap a computation that computes a derivative and delays that work until it is needed, either via the derivative being added to something or being [`unthunk`](@ref)ed manually.
+Thus saving time if it is never used.
```

Comment:
Also combine this with the sentence above

Reply:
```suggestion
thus saving time if it is never used.
```